temporal_sqlalchemy/session.py

```diff
@@ -27,7 +31,54 @@ def persist_history(session: orm.Session, flush_context, instances):
         obj.temporal_options.record_history(obj, session, correlate_timestamp)
 
 
-def temporal_session(session: typing.Union[orm.Session, orm.sessionmaker], strict_mode=False) -> orm.Session:
+def persist_history_on_commit(session: orm.Session, flush_context, instances):
+    if any(_temporal_models(session.deleted)):
+        raise ValueError("Cannot delete temporal objects.")
+
+    metadata = get_session_metadata(session)
+
+    correlate_timestamp = datetime.datetime.now(tz=datetime.timezone.utc)
+    for obj in _temporal_models(itertools.chain(session.dirty, session.new)):
+        if obj.temporal_options.allow_persist_on_commit:
+            metadata[CHANGESET_KEY].add(obj)
+        else:
+            obj.temporal_options.record_history(obj, session, correlate_timestamp)
+
+    if metadata[IS_COMMITTING_KEY]:
+        for obj in metadata[CHANGESET_KEY]:
+            obj.temporal_options.record_history(obj, session, correlate_timestamp)
+
+        metadata[CHANGESET_KEY].clear()
+        metadata[IS_COMMITTING_KEY] = False
+
+
+def enable_is_committing_flag(session):
+    metadata = get_session_metadata(session)
+    metadata[IS_COMMITTING_KEY] = True
+
+
+PERSIST_ON_COMMIT_LISTENERS = (
+    ('before_flush', persist_history_on_commit),
+    ('before_commit', enable_is_committing_flag),
```

Comment:
`before_commit` happens before the final `before_flush`, so we have to set a flag that does the operation.  It's really not ideal, but this is the order of the lifecycle events.